[feat] ppl

wu-hui opened the pull request.

wu-hui committed:
introduce pipeline proto changes

wu-hui committed:
Add expressions, stages, and pipeline

wu-hui committed:
can run pipelines

wu-hui committed:
pipeline impl

MarkDuckworth commented on dev/system-test/pipeline.ts:
```diff
@@ -0,0 +1,733 @@
+  it('arrayContainsAll works', async () => {
+    const results = await randomCol
+      .pipeline()
+      .where(Field.of('tags').arrayContainsAll('adventure', 'magic'))
```
I wish this read `.where(field('tags').arrayContains...` or even `.where(Field.named('tags').arrayContains...`. As I read through the code, I continue to read it as "where field of something" but the meaning is "where field something".

ehsannas replied:
I think for Java Andy added an function overloads that covers the "common" usecase -- which is the case that the first parameter is a string specifying the name of the field. Is that possible to do for TS?

MarkDuckworth commented on dev/system-test/pipeline.ts:
```diff
@@ -0,0 +1,733 @@
+    const results = await randomCol
+      .pipeline()
+      .select(
+        Field.of('awards').mapGet('hugo').as('hugoAward'),
```
I'd also consider `Field.of('awards.hugo').as('hugoAward')`, or is there some special behavior to `mapGet`?

Does `mapGet()` support dot notation? 
`Field.of('awards').mapGet('hugo.year').as('hugoAwardYear')`
vs
`Field.of('awards').mapGet('hugo').mapGet('year').as('hugoAwardYear')`

wu-hui replied:
Good question. It looks like mapGet does not support nested fields (always returns null). 

One benefit is maybe mapGet supports field with special chars better (., /, etc).

MarkDuckworth commented on dev/system-test/pipeline.ts:
```diff
@@ -0,0 +1,733 @@
+    const results = await randomCol
+      .pipeline()
+      .select(
+        cosineDistance(Constant.ofVector(sourceVector), targetVector).as(
```
Consider just `Constant.vector(sourceVector)`

MarkDuckworth commented on dev/system-test/pipeline.ts:
```diff
@@ -0,0 +1,733 @@
+        ),
+        euclideanDistance(Constant.ofVector(sourceVector), targetVector).as(
+          'euclideanDistance'
+        )
```
This pattern will support KNN. If eventually there is ANN support we may need a different pattern. Do we need to consider that now?

MarkDuckworth commented on dev/system-test/pipeline.ts:
```diff
@@ -0,0 +1,733 @@
+  it('testNestedFields', async () => {
+    const results = await randomCol
+      .pipeline()
+      .where(eq('awards.hugo', true))
```
Cool. Does this make `mapGet` redundant?

wu-hui replied:
See above.

MarkDuckworth reviewed (commented):
Just starting review. Will resume next week, but providing some thoughts now.

MarkDuckworth commented on dev/src/stage.ts:
```diff
@@ -0,0 +1,245 @@
+  }
+}
+
+export class GenerateStage implements Stage {
```
Should this be `GenericStage` to match the API method `Pipeline.genericStage()`

MarkDuckworth commented on dev/src/pipeline.ts:
```diff
@@ -0,0 +1,612 @@
+  }
+}
+
+export class PipelineResult<
```
The RPC API appears to call these `Document`s

MarkDuckworth commented on dev/src/expression.ts:
```diff
@@ -0,0 +1,1425 @@
+
+  arrayConcat(...values: Expr[]): ArrayConcat;
+  arrayConcat(...values: any[]): ArrayConcat;
+  arrayConcat(...values: any[]): ArrayConcat {
```
Consider making values a standard array parameter, not rest parameters.

MarkDuckworth commented on dev/src/expression.ts:
```diff
@@ -0,0 +1,1425 @@
+  dotProductDistance(other: Expr): DotProductDistance;
+  dotProductDistance(other: VectorValue): DotProductDistance;
+  dotProductDistance(other: number[]): DotProductDistance;
+  dotProductDistance(other: Expr | VectorValue | number[]): DotProductDistance {
```
Might be more accurate to just call this this and the class `dotProduct`. Drop "Distance"

wu-hui replied:
Will do.

MarkDuckworth commented on dev/src/expression.ts:
```diff
@@ -0,0 +1,1425 @@
+  static of(value: Array<any>): Constant;
+  static of(value: Map<string, any>): Constant;
+  static of(value: VectorValue): Constant;
+  static of(value: api.IValue): Constant;
```
If you want to make this more extensible, you could also add an interface

```
interface FirestoreProtoSerializable {
  toProto(): api.IValue;
}
```

and have a `static of(value: FirestoreProtoSerializable): Constant`

MarkDuckworth commented on dev/src/expression.ts:
```diff
@@ -0,0 +1,1425 @@
+  }
+}
+
+class DotProductDistance extends Function {
```
Same. Just "DotProduct" is probably more accurate.

MarkDuckworth commented on dev/src/expression.ts:
```diff
@@ -0,0 +1,1425 @@
+    return new Fields([Field.of(name), ...others.map(Field.of)]);
+  }
+
+  static ofAll(): Fields {
```
Is this used?

wu-hui replied:
Will remove.

MarkDuckworth commented on dev/src/expression.ts:
```diff
@@ -0,0 +1,1425 @@
+  filterable = true as const;
+}
+
+class ArrayContainsAny extends Function implements FilterCondition {
```
Is FilterCondition always a function that returns a boolean? If so, would it be useful to make a more generic BooleanFunction interface in case there is a future use case for boolean functions outside of filtering?

MarkDuckworth commented on dev/src/expression.ts:
```diff
@@ -0,0 +1,1425 @@
+export function add(left: string, right: Expr): Add;
+export function add(left: string, right: any): Add;
+export function add(left: Expr | string, right: Expr | any): Add {
+  const normalizedLeft = typeof left === 'string' ? Field.of(left) : left;
```
Great candidate for a helper `Field.fromArgument(left)`. Similar to `FieldPath.fromArgument(stringOrFieldPath)`

wu-hui replied:
Will do.

MarkDuckworth commented on dev/src/expression.ts:
```diff
@@ -0,0 +1,1425 @@
+export function add(left: string, right: any): Add;
+export function add(left: Expr | string, right: Expr | any): Add {
+  const normalizedLeft = typeof left === 'string' ? Field.of(left) : left;
+  const normalizedRight = right instanceof Expr ? right : Constant.of(right);
```
`Expr.fromExpressionOrConstant(right)`

wu-hui replied:
Will do.

MarkDuckworth commented on dev/src/expression.ts:
```diff
@@ -0,0 +1,1425 @@
+export function inAny(element: Expr, others: Expr[]): In;
+export function inAny(element: Expr, others: any[]): In;
+export function inAny(element: string, others: Expr[]): In; // Added overload
+export function inAny(element: string, others: any[]): In; // Added overload
```
If this behavior is different from the classic `in()` supporting array, should we also change `Expr.in(...)` to name `Expr.inAny(...)`. Or if this is only supporting `in` arrays, would it make more sense to name it `inArray(...)`?

wu-hui replied:
Will rename Expr.in to inAny.

MarkDuckworth commented on dev/src/expression.ts:
```diff
@@ -0,0 +1,1425 @@
+export function notInAny(element: Expr, others: Expr[]): Not;
+export function notInAny(element: Expr, others: any[]): Not;
+export function notInAny(element: string, others: Expr[]): Not; // Added overload
+export function notInAny(element: string, others: any[]): Not; // Added overload
```
Same

MarkDuckworth reviewed (commented):
Done with reviews for now

wu-hui committed:
Fix routing headers

wu-hui committed:
Basic initial converter support

wu-hui committed:
Converter cleanup

wu-hui committed:
Address some feedbacks and add more nested field test.

wu-hui reviewed (commented):
Thanks for the feedback! I addressed some of those, and maybe we should discuss for the rest.

wu-hui committed:
License header and tests fix

wu-hui committed:
License header and tests fix

wu-hui committed:
Merge remote-tracking branch 'origin/wuandy/PplPP' into wuandy/PplPP

wu-hui commented on dev/system-test/pipeline.ts:
```diff
@@ -0,0 +1,733 @@
+    const results = await randomCol
+      .pipeline()
+      .select(
+        Field.of('tags').arrayConcat('newTag1', 'newTag2').as('modifiedTags')
```
Will do.

wu-hui commented on dev/src/pipeline.ts:
```diff
@@ -0,0 +1,612 @@
+  }
+}
+
+class PaginatingPipeline {
```
Will delete.

wu-hui committed:
Addressing comments.

wu-hui committed:
DTS + some other fixes.

wu-hui committed:
FindNearest update

wu-hui committed:
pipeline and stage dts fix done. expr to go.

wu-hui committed:
fixing expression.ts type decl

wu-hui committed:
length to strLength

wu-hui committed:
remove __path__ hack

wu-hui committed:
resync with expressions catelog

wu-hui committed:
comment out bit operations

wu-hui committed:
more syncs

wu-hui committed:
Fix rest connection by correct url endpoint.

wu-hui committed:
documentation fix

wu-hui committed:
Fix system tests build error by not using `instanceof firestore.Expr` in expression.ts

wu-hui committed:
remove __path__

wu-hui committed:
add reference test

wu-hui committed:
added transaction support.

wu-hui committed:
Add comments

MarkDuckworth commented on dev/src/expression.ts:
```diff
@@ -0,0 +1,6739 @@
+   */
+  static of(name: string): Field;
+  static of(path: firestore.FieldPath): Field;
+  static of(nameOrPath: string | firestore.FieldPath): Field;
```
This overload is redundant. `static of(nameOrPath: string | firestore.FieldPath): Field;` Either remove this or keep it and remove the two above.

MarkDuckworth commented on dev/src/expression.ts:
```diff
@@ -0,0 +1,6739 @@
+   * @param value The Firestore proto value.
+   * @return A new `Constant` instance.
+   */
+  static of(value: api.IValue): Constant;
```
This appears to be redeclared below on line 2083

MarkDuckworth commented on dev/src/expression.ts:
```diff
@@ -0,0 +1,6739 @@
+      return this.value;
+    }
+
+    return serializer.encodeValue(this.value)!;
```
`serializer.endcodeValue(undefined)` will return `null`, which is an permitted, but unhandled case. Do we want to treat undefined as null?

wu-hui committed:
add remove_fields support

wu-hui committed:
add d.ts change for removeFields

MarkDuckworth committed:
fix build

MarkDuckworth committed:
Rename cond, *minimum, *maximum, eqAny, and notEqAny. New tests. Implement NotEqAny.

tom-andersen committed:
chore: Merge main and resolve conflicts. (#2240)

* # Conflicts:
#	dev/conformance/runner.ts
#	dev/src/reference/query.ts
#	dev/system-test/firestore.ts
#	types/firestore.d.ts

* fix

tom-andersen committed:
feature: New Pipeline Stages (#2236)

* new stages

* add tests

* fix

* fix

* Fix code examples

* Add unnest options to `d.ts` file.

* Fix union test

* Add test, fix documentation, refactor.

* Documentation

* Undo

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md
